File: TileCalorimeter/TileMonitoring/python/TileTBBeamMonitorAlgorithm.py
#
#  Copyright (C) 2002-2023 CERN for the benefit of the ATLAS collaboration
#
'''
@file TileTBBeamMonitorAlgorithm.py
@brief Python configuration of TileTBBeamMonitorAlgorithm algorithm for the Run III
'''

from AthenaConfiguration.Enums import Format
from AthenaConfiguration.ComponentFactory import CompFactory


def TileTBBeamMonitoringConfig(flags, fragIDs=[0x100,0x101,0x200,0x201,0x402], **kwargs):

    ''' Function to configure TileTBBeamMonitorAlgorithm algorithm in the monitoring system.'''

    from AthenaConfiguration.ComponentAccumulator import ComponentAccumulator
    result = ComponentAccumulator()

    from TileGeoModel.TileGMConfig import TileGMCfg
    result.merge(TileGMCfg(flags))

    from TileConditions.TileCablingSvcConfig import TileCablingSvcCfg
    result.merge(TileCablingSvcCfg(flags))

    from AthenaMonitoring import AthMonitorCfgHelper
    helper = AthMonitorCfgHelper(flags, 'TileTBBeamMonitoring')

    from AthenaConfiguration.ComponentFactory import CompFactory
    tileTBBeamMonAlg = helper.addAlgorithm(CompFactory.TileTBBeamMonitorAlgorithm, 'TileTBBeamMonAlg')

    tileTBBeamMonAlg.TriggerChain = ''

    kwargs.setdefault('CaloCellContainer', 'AllCalo')
    cellContainer = kwargs['CaloCellContainer']

    kwargs.setdefault('MaskMuonPMTs', [7])

    from TileCalibBlobObjs.Classes import TileCalibUtils as Tile

    modules = []
    if fragIDs:
        for fragID in fragIDs:
            ros = fragID >> 8
            drawer = fragID & 0x3F
            modules += [Tile.getDrawerString(ros, drawer)]
    else:
        for ros in range(1, Tile.MAX_ROS):
            for drawer in range(0, Tile.MAX_DRAWER):
                fragIDs += [(ros << 8) | drawer]
                modules += [Tile.getDrawerString(ros, drawer)]

    tileTBBeamMonAlg.TileFragIDs = fragIDs

    for k, v in kwargs.items():
        setattr(tileTBBeamMonAlg, k, v)

    run = str(flags.Input.RunNumber[0])

    # Configure histogram with TileTBBeamMonAlg algorithm execution time
    executeTimeGroup = helper.addGroup(tileTBBeamMonAlg, 'TileTBBeamMonExecuteTime', 'TestBeam')
    executeTimeGroup.defineHistogram('TIME_execute', path='BeamElements', type='TH1F',
                                     title='Time for execute TileTBBeamMonAlg algorithm;time [#mus]',
                                     xbins=100, xmin=0, xmax=10000)

    nMuonWallPMT = 12
    muonWallPMTArray = helper.addArray([nMuonWallPMT], tileTBBeamMonAlg, 'MuonWallPMT', topPath='TestBeam')
    for postfix, tool in muonWallPMTArray.Tools.items():
        pmt = int(postfix[1:]) + 1
        title = f'Run {run}: Muon Wall PMT{pmt} Amplitude;Amplitude [ADC];Counts'
        name = f'amplitude;MuonWallPMT{pmt}'
        tool.defineHistogram(name, title=title, path='BeamElements', type='TH1F',
                             xbins=410, xmin=0, xmax=4096)

    totalMuEnergyGroup = helper.addGroup(tileTBBeamMonAlg, 'TileTBTotalMuonEnergy', 'TestBeam')
    totalMuEnergyGroup.defineHistogram('TotalMuonEnergy', path='BeamElements', type='TH1F',
                                       title=f'Run {run}: Muon Wall Total Energy; [ADC]',
                                       xbins=1500, xmin=0, xmax=10000)

    nScounters = 3
    sCountersArray = helper.addArray([nScounters], tileTBBeamMonAlg, 'Scounter', topPath='TestBeam')
    for postfix, tool in sCountersArray.Tools.items():
        counter = int(postfix[1:]) + 1
        title = f'Run {run}: S{counter} Counter Amplitude;Amplitude [ADC];Counts'
        name = f'amplitude;S{counter}hist'
        tool.defineHistogram(name, title=title, path='BeamElements', type='TH1F',
                             xbins=410, xmin=0, xmax=4096)

    nCherenkov = 3
    cherenkovArray = helper.addArray([nCherenkov], tileTBBeamMonAlg, 'Cherenkov', topPath='TestBeam')
    for postfix, tool in cherenkovArray.Tools.items():
        cherenkov = int(postfix[1:]) + 1
        title = f'Run {run}: Cherenkov {cherenkov} Amplitude;Amplitude [ADC];Counts'
        name = f'amplitude;Cher{cherenkov}hist'
        tool.defineHistogram(name, title=title, path='BeamElements', type='TH1F',
                             xbins=410, xmin=0, xmax=4096)


    cherCompGroup = helper.addGroup(tileTBBeamMonAlg, 'CherCompare', 'TestBeam')
    cherCompGroup.defineHistogram('amplitude1,amplitude2;CherCompare', path='BeamElements', type='TH2F',
                                  title=f'Run {run}: Cherenkov2 vs Cherenkov1;Amplitude [ADC];Amplitude [ADC]',
                                  xbins=410, xmin=0, xmax=4096, ybins=410, ymin=0, ymax=4096)

    nTOF = 3
    tofArray = helper.addArray([nTOF], tileTBBeamMonAlg, 'TOF', topPath='TestBeam')
    for postfix, tool in tofArray.Tools.items():
        tof = int(postfix[1:]) + 1
        title = f'Run {run}: TOF{tof};[ADC];Counts'
        name = f'amplitude;TOF{tof}'
        tool.defineHistogram(name, title=title, path='BeamElements', type='TH1F',
                             xbins=4096, xmin=-0.5, xmax=4095.5)

    tofDiffGroup = helper.addGroup(tileTBBeamMonAlg, 'TOFDiff', 'TestBeam')
    for tofs in [[2, 1], [2, 3], [3, 1]]:
        title = f'Run {run}: TOF{tofs[0]} - TOF{tofs[1]} Amplitude difference;[ADC];Counts'
        tofDiffGroup.defineHistogram(f'TOFDiff{tofs[0]}{tofs[1]}', title=title, path='BeamElements',
                                     type='TH1F', xbins=4096, xmin=-0.5, xmax=4095.5)


    cherenkovVsTOFArray = helper.addArray([nCherenkov, nTOF], tileTBBeamMonAlg, 'CherenkovVsTOF', topPath='TestBeam')
    for postfix, tool in cherenkovVsTOFArray.Tools.items():
        cherenkovTof = postfix.split('_')
        tof = int(cherenkovTof.pop()) + 1
        cherenkov = int(cherenkovTof.pop()) + 1
        title = f'Run {run}: Cherenkov {cherenkov} Amplitude vs TOF{tof} Amplitude'
        title += f';TOF{tof} Amplitude [ADC];Cherenkov{cherenkov} Amplitude [ADC]'
        name = f'amplitudeTOF,amplitudeCherenkov;Cher{cherenkov}TOF{tof}'
        tool.defineHistogram(name, title=title, path='BeamElements', type='TH2F',
                             xbins=410, xmin=0, xmax=4096, ybins=410, ymin=0, ymax=4096)

    pmtHitMapGroup = helper.addGroup(tileTBBeamMonAlg, 'PMTHitMap', 'TestBeam')
    pmtHitMapGroup.defineHistogram('column,row,amplitude;PMTHitMap', path='BeamElements', type='TProfile2D',
                                   title=f'Run {run}: Muon Wall PMT Hit Map',
                                   xbins=4, xmin=0, xmax=4, ybins=2, ymin=0, ymax=2)

    nScaler = 3
    scalerArray = helper.addArray([nScaler], tileTBBeamMonAlg, 'Scaler', topPath='TestBeam')
    for postfix, tool in scalerArray.Tools.items():
        scaler = int(postfix[1:]) + 1
        title = f'Run {run}: Scaler S{tof};Counts;# Events'
        name = f'counts;Scaler{scaler}'
        tool.defineHistogram(name, title=title, path='BeamElements', type='TH1F',
                             xbins=20000, xmin=-0.5, xmax=19999.5)

    scalerCoincedenceGroup = helper.addGroup(tileTBBeamMonAlg, 'Scaler12', 'TestBeam')
    scalerCoincedenceGroup.defineHistogram('counts12;Scaler12', path='BeamElements', type='TH1F',
                                           title=f'Run {run}: Scaler S1 and S2 coincedence;counts;# Events',
                                           xbins=20000, xmin=-0.5, xmax=19999.5)


    beamChambers = ['BC1', 'BC2']
    beamChambersArray = helper.addArray([beamChambers], tileTBBeamMonAlg, 'BeamChamber', topPath='TestBeam')
    for postfix, tool in beamChambersArray.Tools.items():
        beamChamber = postfix[1:]

        for coordinate in ['X', 'Y']:
            title = f'Run {run}: {beamChamber}{coordinate} Coordinate;{coordinate}[mm];Counts'
            name = f'{beamChamber}{coordinate};{beamChamber}{coordinate}hist'
            tool.defineHistogram(name, title=title, path='BeamElements', type='TH1F',
                                 xbins=201, xmin=-100.5, xmax=100.5)

        tool.defineHistogram(f'{beamChamber}X,{beamChamber}Y;{beamChamber}Profile', path='BeamElements',
                             type='TH2D', title=f'Run {run}: {beamChamber} Beam Profile;X[mm];Y[mm]',
                             xbins=1000, xmin=-100, xmax=100, ybins=1000, ymin=-100, ymax=100)

        tool.defineHistogram(f'{beamChamber}Xsum,{beamChamber}Ysum;{beamChamber}ProfileSum', path='BeamElements',
                             type='TH2D', title=f'Run {run}: {beamChamber} Beam Profile Sum;X [mm];Y [mm]',
                             xbins=1000, xmin=-300, xmax=0, ybins=1000, ymin=-300, ymax=0)


    impactProfileGroup = helper.addGroup(tileTBBeamMonAlg, 'ImpactProfile', 'TestBeam')
    impactProfileGroup.defineHistogram('Ximp,Yimp;ImpactProfile', path='BeamElements', type='TH2F',
                                       title=f'Run {run}: {beamChamber} Impact Profile;X [mm];Y [mm]',
                                       xbins=200, xmin=-100, xmax=100, ybins=200, ymin=-100, ymax=100)


    if cellContainer:
        cherenkovEnergyArray = helper.addArray([nCherenkov], tileTBBeamMonAlg, 'CherenkovVsEnergy', topPath='TestBeam')
        for postfix, tool in cherenkovEnergyArray.Tools.items():
            cherenkov = int(postfix[1:]) + 1
            title = f'Run {run}: Cherenkov {cherenkov} Amplitude vs Total Energy;Energy [pC]; Amplitude [ADC]'
            name = f'totalEnergy,amplitude;Cher{cherenkov}Energy'
            tool.defineHistogram(name, title=title, path='BeamElements', type='TH2F',
                                 xbins=150, xmin=0, xmax=150, ybins=410, ymin=0, ymax=4096)

        cellEneXimpGroup = helper.addGroup(tileTBBeamMonAlg, 'CellEnergyImpactX', 'TestBeam')
        cellEneXimpGroup.defineHistogram('Ximp,cellEnergy;CellEnergyImpactX', path='BeamElements', type='TH2F',
                                         title=f'Run {run}:  Maximum Cell Energy vs Impact X;X [mm];Cell Energy [pc]',
                                         xbins=200, xmin=-100, xmax=100, ybins=150, ymin=0, ymax=150)

        cellEneYimpGroup = helper.addGroup(tileTBBeamMonAlg, 'CellEnergyImpactY', 'TestBeam')
        cellEneYimpGroup.defineHistogram('Yimp,cellEnergy;CellEnergyImpactY', path='BeamElements', type='TH2F',
                                         title=f'Run {run}:  Maximum Cell Energy vs Impact Y;Y [mm];Cell Energy [pc]',
                                         xbins=200, xmin=-100, xmax=100, ybins=150, ymin=0, ymax=150)

        totalEneXimpGroup = helper.addGroup(tileTBBeamMonAlg, 'TotalEnergyImpactX', 'TestBeam')
        totalEneXimpGroup.defineHistogram('Ximp,totalEnergy;TotalEnergyImpactX', path='BeamElements', type='TH2F',
                                          title=f'Run {run}: Total Energy vs Impact X;X [mm];Total Energy [pc]',
                                          xbins=200, xmin=-100, xmax=100, ybins=150, ymin=0, ymax=150)

        totalEneYimpGroup = helper.addGroup(tileTBBeamMonAlg, 'TotalEnergyImpactY', 'TestBeam')
        totalEneYimpGroup.defineHistogram('Yimp,totalEnergy;TotalEnergyImpactY', path='BeamElements', type='TH2F',
                                          title=f'Run {run}: Total Energy vs Impact Y;Y [mm];Total Energy [pc]',
                                          xbins=200, xmin=-100, xmax=100, ybins=150, ymin=0, ymax=150)


        totalEneS1Group = helper.addGroup(tileTBBeamMonAlg, 'ScinCalEnergy', 'TestBeam')
        totalEneS1Group.defineHistogram('amplitude,totalEnergy;ScinCalEnergy', path='BeamElements', type='TH2F',
                                        title=f'Run {run}: Total Energy vs S1 amplitude; Amplitude [ADC];Total Energy [pc]',
                                        xbins=410, xmin=0, xmax=4096, ybins=150, ymin=0, ymax=150)


    accumalator = helper.result()
    result.merge(accumalator)
    return result


if __name__=='__main__':

    # Setup logs
    from AthenaCommon.Logging import log
    from AthenaCommon.Constants import INFO
    log.setLevel(INFO)

    # Set the Athena configuration flags
    from AthenaConfiguration.AllConfigFlags import initConfigFlags
    from AthenaConfiguration.TestDefaults import defaultTestFiles

    flags = initConfigFlags()
    parser = flags.getArgumentParser()
    parser.add_argument('--postExec', help='Code to execute after setup')
    parser.add_argument('--digits', default="TileDigitsCnt", help='Tile digits container')
    parser.add_argument('--channels', default="TileRawChannelCnt",
                        help='Tile raw channel container, if empty they will be reconstructed from digits')
    parser.add_argument('--cells', default="AllCalo",
                        help='Calo Cells container, if empty they will be reconstructed from channels')
    parser.add_argument('--frag-ids', dest='fragIDs', nargs="*", default=['0x100', '0x101', '0x200', '0x201','0x402'],
                        help='Tile Frag IDs of modules to be monitored. Empty=ALL')
    parser.add_argument('--demo-cabling', dest='demoCabling', type=int, default=2018, help='Time Demonatrator cabling to be used')
    parser.add_argument('--nsamples', type=int, default=15, help='Number of samples')
    parser.add_argument('--use-sqlite', dest='useSqlite', default='/afs/cern.ch/user/t/tiledemo/public/efmon/condb/tileSqlite.db',
                        help='Providing local SQlite file, conditions constants will be used from it')
    args, _ = parser.parse_known_args()

    fragIDs = [int(fragID, base=16) for fragID in args.fragIDs]

    flags.Input.Files = defaultTestFiles.RAW_RUN2
    flags.Output.HISTFileName = 'TileTBBeamMonitorOutput.root'
    flags.DQ.useTrigger = False
    flags.DQ.enableLumiAccess = False
    flags.Exec.MaxEvents = 3
    flags.Common.isOnline = True

    flags.Tile.doFit = True
    flags.Tile.useDCS = False
    flags.Tile.NoiseFilter = 0
    flags.Tile.correctTime = False
    flags.Tile.correctTimeJumps = False
    flags.Tile.BestPhaseFromCOOL = False
    flags.Tile.doOverflowFit = False

    if args.channels:
        flags.Tile.RawChannelContainer = args.channels

    flags.fillFromArgs(parser=parser)
    flags.lock()

    flags.dump(pattern='Tile.*|Input.*|Exec.*|IOVDb.[D|G].*', evaluate=True)

    # Initialize configuration object, add accumulator, merge, and run.
    from AthenaConfiguration.MainServicesConfig import MainServicesCfg
    cfg = MainServicesCfg(flags)

    rawChannels = args.channels
    cells = args.cells
    if flags.Input.Format is Format.BS:
        cfg.addPublicTool(CompFactory.TileROD_Decoder(fullTileMode=flags.Input.RunNumber[0]))

        from ByteStreamCnvSvc.ByteStreamConfig import ByteStreamReadCfg
        tileTypeNames = [f'TileDigitsContainer/{args.digits}',
                         'TileRawChannelContainer/TileRawChannelCnt',
                         'TileBeamElemContainer/TileBeamElemCnt']
        if flags.Tile.RunType != 'PHY':
            tileTypeNames += ['TileBeamElemContainer/TileBeamElemCnt']
        cfg.merge(ByteStreamReadCfg(flags, type_names=tileTypeNames))
        cfg.getService('ByteStreamCnvSvc').ROD2ROBmap = ['-1']
    else:
        from AthenaPoolCnvSvc.PoolReadConfig import PoolReadCfg
        cfg.merge(PoolReadCfg(flags))

    if not rawChannels:
        # Run reconstruction to produce Tile raw channels
        rawChannels = flags.Tile.RawChannelContainer

        from TileRecUtils.TileRawChannelMakerConfig import TileRawChannelMakerCfg
        cfg.merge(TileRawChannelMakerCfg(flags))
        if args.threads and (args.threads > 1):
            rawChMaker = cfg.getEventAlgo('TileRChMaker')
            rawChMaker.Cardinality = args.threads

        if args.useSqlite:
            cfg.getService('IOVDbSvc').overrideTags += [
                f'<prefix>/TILE</prefix> <db>sqlite://;schema={args.useSqlite};dbname={flags.IOVDb.DatabaseInstance}</db>'
            ]


    if not cells:
        # Run reconstruction to produce Tile cells
        cells = 'AllCalo'
        from TileRecUtils.TileCellMakerConfig import TileCellMakerCfg
        cfg.merge(TileCellMakerCfg(flags, maskBadChannels=False, mergeChannels=False, UseDemoCabling=args.demoCabling))


    cfg.merge(TileTBBeamMonitoringConfig(flags, TBperiod=2021, fragIDs=fragIDs))

    from TileConditions.TileInfoLoaderConfig import TileInfoLoaderCfg
    cfg.merge(TileInfoLoaderCfg(flags))
    tileInfoLoader = cfg.getService('TileInfoLoader')
    tileInfoLoader.NSamples = args.nsamples
    tileInfoLoader.TrigSample = (args.nsamples - 1) // 2  # Floor division

    # Any last things to do?
    if args.postExec:
        log.info('Executing postExec: %s', args.postExec)
        exec(args.postExec)

    cfg.printConfig(withDetails=True, summariseProps=True)

    cfg.store(open('TileTBBeamMonitorAlgorithm.pkl', 'wb'))

    sc = cfg.run()

    import sys
    # Success should be 0
    sys.exit(not sc.isSuccess())

File: TileCalorimeter/TileMonitoring/src/TileTBBeamMonitorAlgorithm.cxx
/*
  Copyright (C) 2002-2023 CERN for the benefit of the ATLAS collaboration
*/

#include "TileTBBeamMonitorAlgorithm.h"
#include "TileCalibBlobObjs/TileCalibUtils.h"
#include "TileIdentifier/TileHWID.h"
#include "TileIdentifier/TileTBFrag.h"

#include "StoreGate/ReadHandle.h"

#include <algorithm>

StatusCode TileTBBeamMonitorAlgorithm::initialize() {

  ATH_MSG_INFO("in initialize()");
  ATH_CHECK( AthMonitorAlgorithm::initialize() );

  ATH_CHECK( m_beamElemContainerKey.initialize() );

  ATH_CHECK( m_cablingSvc.retrieve() );
  ATH_CHECK( detStore()->retrieve(m_tileHWID) );

  std::vector<std::string> modules;
  for (int fragID : m_fragIDs) {
    int ros = fragID >> 8;
    int drawer = fragID & 0x3F;
    modules.push_back(TileCalibUtils::getDrawerString(ros, drawer));
  }

  std::ostringstream os;
  if ( m_fragIDs.size() != 0) {
    std::sort(m_fragIDs.begin(), m_fragIDs.end());
    for (int fragID : m_fragIDs) {
      unsigned int ros    = fragID >> 8;
      unsigned int drawer = fragID & 0xFF;
      std::string module = TileCalibUtils::getDrawerString(ros, drawer);
      os << " " << module << "/0x" << std::hex << fragID << std::dec;
    }
  } else {
    os << "NONE";
  }

  ATH_MSG_INFO("Monitored modules/frag ID:" << os.str());

  using namespace Monitored;

  m_tofGroups = buildToolMap<int>(m_tools, "TOF", N_TOF);
  m_sCounterGroups = buildToolMap<int>(m_tools, "Scounter", N_S_COUNTER);
  m_cherenkovGroups = buildToolMap<int>(m_tools, "Cherenkov", N_CHERENKOV);
  m_muonWallGroups = buildToolMap<int>(m_tools, "MuonWallPMT", N_MUON_WALL_PMT);

  std::vector<std::string> beamChambers{"BC1", "BC2"};
  m_beamChamberGroups = buildToolMap<int>(m_tools, "BeamChamber", beamChambers);

  return StatusCode::SUCCESS;
}


StatusCode TileTBBeamMonitorAlgorithm::fillHistograms( const EventContext& ctx ) const {

  // In case you want to measure the execution time
  auto timer = Monitored::Timer("TIME_execute");

  uint32_t run = GetEventInfo(ctx)->runNumber();

  using Tile = TileCalibUtils;

  // TDC/BEAM Items
  int muTag = 0;
  int muHalo = 0;
  int muVeto = 0;

  std::vector<std::reference_wrapper<int>> mu{muTag, muHalo, muVeto};

  int sCounter[N_S_COUNTER] = {0};
  int cherenkov[N_CHERENKOV] = {0};

  // MUON/MuWall
  float muonWall[N_MUON_WALL_PMT] = {0};

  int tof[16] = {0};
  int btdc[16] = {0};
  int btdcHitsN[16] = {0};

  SG::ReadHandle<TileBeamElemContainer> beamElemContainer(m_beamElemContainerKey, ctx);
  ATH_CHECK( beamElemContainer.isValid() );

  ATH_MSG_VERBOSE("TileRawDataContainer of TileRawDataCollection of TileBeamElem size = " << beamElemContainer->size());

  for (const TileBeamElemCollection* beamElemCollection : *beamElemContainer) {

    // Retreive frag identifiers
    int frag = beamElemCollection->identify();
    ATH_MSG_VERBOSE("TileRawDataCollection of TileBeamElem Id = 0x" << MSG::hex << frag << MSG::dec
                    << " size = " << beamElemCollection->size());

    for (const TileBeamElem* beamElement : *beamElemCollection) {

      ATH_MSG_VERBOSE((std::string) *beamElement);

      HWIdentifier id = beamElement->adc_HWID();
      std::vector<uint32_t> digits = beamElement->get_digits();
      int channel = m_tileHWID->channel(id);
      int nDigits = digits.size();

      if ( nDigits <= 0 ) {
        ATH_MSG_ERROR("Wrong no. of samples (" << nDigits
                      << ") for channel " << channel
                      << " in frag 0x"<< MSG::hex << frag << MSG::dec
                      << " - " << BeamFragName[frag & 0x1F]);
      } else {

        int amplitude = digits[0];

        ATH_MSG_DEBUG("Found channel " << channel
                      << " in frag 0x" << MSG::hex << frag << MSG::dec
                      << " - " << BeamFragName[frag & 0x1F]
                      << " with amp=" << amplitude);

        switch (frag) {

        case MUON_ADC_FRAG:

          if(channel >= 0 && channel < 8) {
            muonWall[channel] = amplitude;
          } else {
            errorWrongChannel(frag, channel);
          }

          break;

        case ADDR_ADC_FRAG:

           if(channel >= 0 && channel < 6) {
            muonWall[8 + channel] = amplitude;
          } else {
            errorWrongChannel(frag, channel);
          }

          break;

        case COMMON_TOF_FRAG:
          if (m_TBperiod >= 2022) {
            // The first channels are connected to BC1 and BC2, the last 4 channels are supposed to be TOF
            if (channel > 11) {
              if(channel < 16) {
                tof[channel] = amplitude;
                ATH_MSG_VERBOSE( "TOF: " << channel << " amp: " << amplitude);
              } else {
                errorWrongChannel(frag, channel);
              }
              break;
            }
            // Fall through to case COMMON_TDC1_FRAG to unpack the first channels of BC1 and BC2
            [[fallthrough]]; // silent the warning on fall through
          } else {
            break;
          }
        case COMMON_TDC1_FRAG:

          if ((channel > 11) && (channel < 16) && (run > 2211136)) {
            tof[channel] = amplitude;
            ATH_MSG_VERBOSE( "TOF: " << channel << " amp: " << amplitude);
          } if (channel < 16) {
            if (m_TBperiod >= 2021) {
              if (btdcHitsN[channel] == 0) {
                btdc[channel] = amplitude;
                ++(btdcHitsN[channel]);
              }
            } else {
              btdc[channel] = amplitude;
            }
          } else errorWrongChannel(frag, channel);
          break;

        case BEAM_ADC_FRAG:

          if (channel >= 0 && channel < 8) {
            if (channel < 3) {
              sCounter[channel] = amplitude;
            } else if (channel < 5) {
              cherenkov[channel - 3] = amplitude;
            } else {
              mu[channel - 5] = amplitude;
            }
          } else {
            errorWrongChannel(frag, channel);
          }


        case COMMON_ADC1_FRAG:

          if (run > 2211444) {

            if (channel >= 0 && channel < 16) {
              if (channel < 2) {
                sCounter[channel] = amplitude;
              } else if (channel == 2) {
                muonWall[10] = amplitude;
              } else if (channel < 6) {
                cherenkov[channel - 3] = amplitude;
              } else {
                muonWall[channel - 6] = amplitude;
              }
            } else {
              errorWrongChannel(frag, channel);
            }

            break;

          } else {

            if (channel >= 0 && channel < 16) {
              if (channel < 3) {
                sCounter[channel] = amplitude;
              } else if (channel < 6) {
                cherenkov[channel - 3] = amplitude;
              }
            } else {
              errorWrongChannel(frag, channel);
            }

            break;
          }


        case COMMON_ADC2_FRAG:

          if (run < 2211445) {

            if(channel >= 0 && channel < 16) {
              muonWall[channel] = amplitude;
            } else if (channel > 31) {
              errorWrongChannel(frag, channel);
            }
          }
          break;
        }
      }
    }
  }

  for (int counter = 0; counter < N_TOF; ++counter) {
    std::vector<int> counterToTOF{14,15,13};
    auto monAmplitude = Monitored::Scalar<double>("amplitude", tof[counterToTOF[counter]]);
    fill(m_tools[m_tofGroups[counter]], monAmplitude);
  }

  for (int counter = 0; counter < N_S_COUNTER; ++counter) {
    auto monAmplitude = Monitored::Scalar<double>("amplitude", sCounter[counter]);
    fill(m_tools[m_sCounterGroups[counter]], monAmplitude);
  }

  for (int counter = 0; counter < N_CHERENKOV; ++counter) {
    auto monAmplitude = Monitored::Scalar<double>("amplitude", cherenkov[counter]);
    fill(m_tools[m_cherenkovGroups[counter]], monAmplitude);
  }

  for (int pmt = 0; pmt < N_MUON_WALL_PMT; ++pmt) {
    auto monAmplitude = Monitored::Scalar<double>("amplitude", muonWall[pmt]);
    fill(m_tools[m_muonWallGroups[pmt]], monAmplitude);
  }

  for (int row = 0; row < 2; ++row) {
    for (int column = 0; column < 4; ++column) {
      auto monRow = Monitored::Scalar<double>("row", row);
      auto monColumn = Monitored::Scalar<double>("column", column);
      auto monAmplitude = Monitored::Scalar<double>("amplitude", muonWall[8 - (row * 4 + column)]);
      fill("PMTHitMap", monColumn, monRow, monAmplitude);
    }
  }


  // Beam Chamber Coordinates
  // For BC1
  auto bc1X = Monitored::Scalar<double>("BC1X", 0.);
  auto bc1Y = Monitored::Scalar<double>("BC1Y", 0.);
  if (run > 2211444) {
    bc1X = (btdc[8] - btdc[0]) * m_bc1HorizontalSlope + m_bc1HorizontalOffset; // (right - left)
    bc1Y = (btdc[9] - btdc[3]) * m_bc1VerticalSlope + m_bc1VerticalOffset; // (up - down)
  } else {
    bc1X = (btdc[1] - btdc[0]) * m_bc1HorizontalSlope + m_bc1HorizontalOffset;  // (right - left)
    bc1Y = (btdc[2] - btdc[3]) * m_bc1VerticalSlope + m_bc1VerticalOffset; // (up - down)
  }
  fill(m_tools[m_beamChamberGroups.at("BC1")], bc1X);
  fill(m_tools[m_beamChamberGroups.at("BC1")], bc1Y);
  fill(m_tools[m_beamChamberGroups.at("BC1")], bc1X, bc1Y);

  // For BC2:
  auto bc2X = Monitored::Scalar<double>("BC2X", (btdc[5] - btdc[4]) * m_bc2HorizontalSlope + m_bc2HorizontalOffset); // (right - left)
  auto bc2Y = Monitored::Scalar<double>("BC2Y", (btdc[6] - btdc[7]) * m_bc2VerticalSlope + m_bc2VerticalOffset); // (up - down)
  fill(m_tools[m_beamChamberGroups.at("BC2")], bc2X);
  fill(m_tools[m_beamChamberGroups.at("BC2")], bc2Y);
  fill(m_tools[m_beamChamberGroups.at("BC2")], bc2X, bc2Y);

  // Sum Plots
  // For BC1
  auto bc1Xsum = Monitored::Scalar<double>("BC1Xsum", 0.);
  auto bc1Ysum = Monitored::Scalar<double>("BC1Ysum", 0.);;
  if (run > 2211444) {
    bc1Xsum =(btdc[8] + btdc[0]) * m_bc1HorizontalSlope + m_bc1HorizontalOffset;
    bc1Ysum = (btdc[9] + btdc[3]) * m_bc1VerticalSlope + m_bc1VerticalOffset;
  } else {
    bc1Xsum = (btdc[1] + btdc[0]) * m_bc1HorizontalSlope + m_bc1HorizontalOffset;
    bc1Ysum = (btdc[2] + btdc[3]) * m_bc1VerticalSlope + m_bc1VerticalOffset;
  }
  fill(m_tools[m_beamChamberGroups.at("BC1")], bc1Xsum, bc1Ysum);

  //For BC2
  auto bc2Xsum = Monitored::Scalar<double>("BC2Xsum", (btdc[5] + btdc[4]) * m_bc2HorizontalSlope + m_bc2HorizontalOffset);
  auto bc2Ysum = Monitored::Scalar<double>("BC2Ysum", (btdc[6] + btdc[7]) * m_bc2VerticalSlope + m_bc2VerticalOffset);
  fill(m_tools[m_beamChamberGroups.at("BC2")], bc2Xsum, bc2Ysum);

  //Impact Coordinates
  // For one cell, plot energy total as a function of x Impact -- (xImp, cell_energy)...
  auto xImp = Monitored::Scalar<double>("Ximp", bc2X + (bc2X - bc1X) * m_beamBC2Z / (m_beamBC1Z - m_beamBC2Z));
  auto yImp = Monitored::Scalar<double>("Yimp", bc2Y + (bc2Y - bc1Y) * m_beamBC2Z / (m_beamBC1Z - m_beamBC2Z));
  fill("ImpactProfile", xImp, yImp);

  fill("TileTBBeamMonExecuteTime", timer);

  return StatusCode::SUCCESS;
}


void TileTBBeamMonitorAlgorithm::errorWrongChannel(int frag, int channel) const {
  ATH_MSG_ERROR("Wrong channel " << channel
                << " in frag 0x" << MSG::hex << frag << MSG::dec
                << " - " << BeamFragName[frag & 0x1F]);
}
